ROADMAP.md

```diff
@@ -1,26 +1,193 @@
+
+## __TENTATIVE__: Uniform Resource Model
+
+Right now there is no uniform model for the resources Machine creates (and,
+indeed, sometimes it is obligated to created resources such as security groups
+to ensure that the correct ports are open for connection, SSH keys are created
+properly, and so on).  This causes all sorts of issues due to inconsistency and
+confusion of responsibilty: currently the drivers are expected to clean up all
+of the resources they create, but do not always do so reliably, etc.
+
+The Uniform Resource Model would start small by defining an interface such as
+follows:
+
+```
+type Resource interface {
+        Create() (error)
+        Read() (bool, error) // Can now distinguish between an "error read" and the resource not existing - no direct way of tackling this today
+        Update() (error) // Would be used to change parameters such as instance size when needed - should be able to be called idempotently
+        Delete() (error)
+        Parents() ([]Resource, error)
+}
+```
+
+Instead of invoking `Create`, `Delete`, etc. directly on the instances, they
+would instead be part of a uniform resource creation wrapper so that other,
+dependent resources (a limited subset, since Machine should only account for
+what it absolutely has to) can be created reliably and verified to exist as part
+of creation.  Without such a unifying model, Machine will continue to be plagued
+by "out-of-state" bugs which it has no way of recovering, except by
+brute-forcing checks for the relevant resources in every location.
+
+This is also intended to make it possible to recover from error conditions, e.g.
+something upstream failed during provision, but we still want to be able to
+`Create` our `ProvisionedResource`, which is something that would theoretically
+be possible with this model, since Machine could detect the current state of the
+system and see where on the "map" (declared configuration) to re-attempt from.
+Because we have a very limited sub-set of things we wish Machine to account for
+with respect to provisioning, this seems viable.
+
+The design of such a system will have to be considered carefully, and ideally
+rolled out slowly, so that it can be proven to be stable before moving all of
+the resources over to use it.  We would start by moving all of the `Driver`
+implementations to fulfill this interface, then moving all of the associated
+resources (SSH keys, security groups, etc.) over to the new model as it becomes
+more mature.  As time goes on, properties which can be modified by `Update` can
+be added, until they are supported Machine will simply error out with a message
+saying that updating that parameters is not supported.
+
+This goal is the most tentative of all listed here, but the problems it
+addresses are something which must be tackled in some fashion eventually if
+Machine is to become a reliable, and error-resilient, tool.
+
+## Support for creating multiple instances at once
```

Comment:
maybe move this section up above "Uniform Resource Model" so it's not mistaken as being tentative?